feat: don't return single packed values as sequences

BenjaminMummery opened the pull request:
# The problem

The behavior of wf_run.get_results() differs between runtimes when the workflow function returns a single, packed artifact.

## Example

```python
@sdk.task
def list_():
    return [1, 2, 3]

@sdk.workflow
def wf1():
    a = list_()
    return a
wf_run = wf1().run(runtime)
wf_run.wait_until_finished()
print("results", wf_run.get_results())
print("artifacts", wf_run.get_artifacts())
print("task outputs", [t.get_outputs() for t in wf_run.get_tasks()])
```

In process:
```
results [1, 2, 3]
artifacts {'invocation-0-task-list-': [1, 2, 3]}
task outputs [[1, 2, 3]]
```

Ray:
```
results ([1, 2, 3],)
artifacts {'invocation-0-task-list-': [1, 2, 3]}
task outputs [[1, 2, 3]]
```

CE:
```
results ([1, 2, 3],)
artifacts {'invocation-0-task-list-': ([1, 2, 3],)}
task outputs [([1, 2, 3],)]
```

QE:
```
results ([1, 2, 3],)
artifacts {'invocation-0-task-list-': [1, 2, 3]}
task outputs [[1, 2, 3]]
```

# This PR's solution

The API, and in particular `WorkflowRun.get_results()`, behaves identically to vanilla python for all runtimes - whatever you would get back from running your workflow as normal python functions is what you get back from `get_results`. In practice, this means that 2 or more results are returned as a tuple, single results are returned as-is.

The CLI checks whether the results are a tuple and, if not, converts them to a 1-element tuple which can then be iterated over identically to an n-element tuple.

# Checklist

_Check that this PR satisfies the following items:_

- [X] Tests have been added for new features/changed behavior (if no new features have been added, check the box).
- [ ] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [X] The PR's title is prefixed with `<feat/fix/chore/internal/docs>[!]:`

BenjaminMummery committed:
test: update tests for packed single returns.

[ORQSDK-801]

BenjaminMummery committed:
test: add tests for consistent returns across runtimes.

[ORQSDK-801]

BenjaminMummery committed:
feat: return single results without wrapping them in a sequence.

[ORQSDK-801]

BenjaminMummery committed:
refactor: unify workflow definitions for consistency tests.

[ORQSDK-801]

BenjaminMummery committed:
chore: remove outdated todo.

[ORQSDK-801]

alexjuda commented on src/orquestra/sdk/_base/_api/_wf_run.py:
```diff
@@ -311,7 +311,7 @@ def get_status_model(self) -> WorkflowRunModel:
         return self._runtime.get_workflow_run_status(run_id)
 
-    def get_results(self, wait: bool = False) -> t.Sequence[t.Any]:
+    def get_results(self, wait: bool = False, force_as_sequence: bool = False) -> t.Any:
```
IMO the multi-output behavior was already difficult to reason about. This option additionally increases the number of cases to consider by 2. Is there a good use case for it?

BenjaminMummery replied:
My reasoning for it went like this: The CLI needs to know how many results you got back in order to format the output properly. At present it uses the len of the return from get_results to determine this. With these changes, if I have 1 result which is an iterable of len n, the cli would incorrectly report this as n results. Therefore we need a way for the CLI to be able to reliably determine the number of results. 

We could do this via private methods, however I could see a use-case where a user has written general-purpose code to handle the results of workflows, and would therefore run into the same ambiguity problem as the CLI. Therefore it made sense to me to have this be a public option. WDYT?

alexjuda replied:
Right, CLI needs to know if it deals with one or multiple values because it affects the formatting. I also agree that CLI should only use public APIs.

I like @jamesclark-Zapata's point that our goal should be to do whatever regular Python code would.

Perhaps we can the problem around? How would need the CLI work if we wanted to implement `orq wf results` for regular Python functions instead of workflows?

jamesclark-Zapata replied:
Could it be as simple as:
```
results = run.get_results()
n_outputs = len(results) if isinstance(results, tuple) else 1
...
```

alexjuda replied:
I like the `if isinstance(x, tuple)`. I would worry about edge cases if we put this under the API implementation but if it's moved to the formatting code at the last layer (CLI) we should be fine.

BenjaminMummery replied:
Okay I like this. As of [f102239](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/141/commits/f102239cf0349008eb95efd68ba26f419cc47589) `force_as_sequence` no longer exists. Instead the API behaves in line with vanilla python, and the CLI converts non-tuple values into 1-element tuples that can then be iterated over just like before.

BenjaminMummery committed:
Test: update test stdout strings for windows.

Windows will print `blah\r\n` where mac or linux would print `blah\n`, so expecting strings that include only the newline will cause tests to fail on windows.

[ORQSDK-801]

BenjaminMummery committed:
refactor: remove

[ORQSDK-801]
The cli now converts non-tuple values into single-value tuples.

BenjaminMummery committed:
style: remove trailing whitespace.

[ORQSDK-801]

BenjaminMummery committed:
test: check API results explicitely against vanilla python.

[ORQSDK-801]

jamesclark-Zapata reviewed (approved):
A few comments, but overall lgtm

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-801-invalid-wf-result-shape-packed-future

BenjaminMummery committed:
refactor: remove unneeded change that separated state and status model calls.

[ORQSDK-801]

BenjaminMummery commented on src/orquestra/sdk/_base/_api/_wf_run.py:
```diff
@@ -347,14 +347,15 @@ def get_results(self, wait: bool = False) -> t.Sequence[t.Any]:
 
-        if (state := self.get_status()) not in COMPLETED_STATES:
+        status_model = self.get_status_model()
+        if (state := status_model.status.state) not in COMPLETED_STATES:
```
It is not, removed in [5c0bc81](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/141/commits/5c0bc8129075eb8a82e09a6c00eef60b213aeac4)

BenjaminMummery committed:
style: update typehint for get_results.

[ORQSDK-801]

BenjaminMummery committed:
fix: remove trailing carrage return from ray id on windows.

[ORQSDK-801]

BenjaminMummery committed:
test: strip carrage returns from stdout before comparison.

[ORQSDK-801]

BenjaminMummery committed:
docs: update changelog.

[ORQSDK-801]

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-801-invalid-wf-result-shape-packed-future

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-801-invalid-wf-result-shape-packed-future
